verifiers/internal/gha/builder.go

```diff
@@ -33,63 +34,81 @@ var defaultContainerTrustedReusableWorkflows = map[string]bool{
+		return nil, fmt.Errorf("%w: workflow uri: %s", serrors.ErrorMalformedURI, id.JobWobWorkflowRef)
 	}
 
-	// Trusted workflow verification by name.
+	// Verify trusted workflow.
 	reusableWorkflowPath := strings.Trim(workflowPath[0], "/")
-	builderID, err := verifyTrustedBuilderID(reusableWorkflowPath,
+	reusableWorkflowTag := strings.Trim(workflowPath[1], "/")
+	builderID, err := verifyTrustedBuilderID(reusableWorkflowPath, reusableWorkflowTag,
 		builderOpts.ExpectedID, defaultBuilders)
 	if err != nil {
-		return "", err
+		return nil, err
 	}
 
-	// Verify the ref.
-	if err := verifyTrustedBuilderRef(id, strings.Trim(workflowPath[1], "/")); err != nil {
-		return "", err
+	// Verify the ref is a full semantic version tag.
+	if err := verifyTrustedBuilderRef(id, reusableWorkflowTag); err != nil {
+		return nil, err
 	}
 
 	// Issuer verification.
 	if !strings.EqualFold(id.Issuer, certOidcIssuer) {
-		return "", fmt.Errorf("untrusted token issuer: %s", id.Issuer)
+		return nil, fmt.Errorf("%w: %s", serrors.ErrorInvalidOIDCIssuer, id.Issuer)
 	}
 
 	// The caller repository in the x509 extension is not fully qualified. It only contains
 	// {org}/{repository}.
 	expectedSource := strings.TrimPrefix(source, "git+https://")
 	expectedSource = strings.TrimPrefix(expectedSource, "github.com/")
 	if !strings.EqualFold(id.CallerRepository, expectedSource) {
-		return "", fmt.Errorf("%w: expected source '%s', got '%s'", serrors.ErrorMismatchSource,
+		return nil, fmt.Errorf("%w: expected source '%s', got '%s'", serrors.ErrorMismatchSource,
 			expectedSource, id.CallerRepository)
 	}
 
 	// Return the builder and its tag.
+	// Note: the tag has the format `refs/tags/v1.2.3`.
 	return builderID, nil
 }
 
 // Verifies the builder ID at path against an expected builderID.
 // If an expected builderID is not provided, uses the defaultBuilders.
-func verifyTrustedBuilderID(path string, builderID *string, defaultBuilders map[string]bool) (string, error) {
+func verifyTrustedBuilderID(certPath, certTag string, expectedBuilderID *string, defaultBuilders map[string]bool) (*utils.BuilderID, error) {
+	var builderID *utils.BuilderID
+	var err error
+	certBuilderName := "https://github.com/" + certPath
+	// WARNING: we don't validate the tag here, because we need to allow
+	// refs/heads/main for e2e tests. See verifyTrustedBuilderRef().
 	// No builder ID provided by user: use the default trusted workflows.
-	if builderID == nil || *builderID == "" {
-		if _, ok := defaultBuilders[path]; !ok {
-			return "", fmt.Errorf("%w: %s got %t", serrors.ErrorUntrustedReusableWorkflow, path, builderID == nil)
+	if expectedBuilderID == nil || *expectedBuilderID == "" {
+		if _, ok := defaultBuilders[certPath]; !ok {
+			return nil, fmt.Errorf("%w: %s got %t", serrors.ErrorUntrustedReusableWorkflow, certPath, expectedBuilderID == nil)
+		}
+		// Construct the builderID using the default builder's name and the certificate's tag.
```

Comment:
this is not the default right? `certPath` is from the signing certiicate

Reply:
+1, change made